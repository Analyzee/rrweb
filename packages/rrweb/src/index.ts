import record from './record';
import { Replayer } from './replay';
import { _mirror } from './utils';
import * as utils from './utils';

export {
  EventType,
  IncrementalSource,
  MouseInteractions,
  ReplayerEvents,
} from '@rrweb/types';

export * from './plugins/network/record';
export * from './plugins/network/replay';

export type { recordOptions } from './types';

const { addCustomEvent } = record;
const { freezePage } = record;

export {
  record,
  addCustomEvent,
  freezePage,
  Replayer,
  _mirror as mirror,
  utils,
};
