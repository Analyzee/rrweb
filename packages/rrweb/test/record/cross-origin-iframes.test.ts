import * as fs from 'fs';
import * as path from 'path';
import type * as puppeteer from 'puppeteer';
import type {
  recordOptions,
  listenerHandler,
  eventWithTime,
  mutationData,
} from '../../src/types';
import { EventType, IncrementalSource } from '../../src/types';
import {
  assertSnapshot,
  getServerURL,
  launchPuppeteer,
  startServer,
  stripBase64,
  waitForRAF,
} from '../utils';
import type * as http from 'http';

interface ISuite {
  code: string;
  browser: puppeteer.Browser;
  page: puppeteer.Page;
  events: eventWithTime[];
  server: http.Server;
  serverURL: string;
}

interface IWindow extends Window {
  rrweb: {
    record: (
      options: recordOptions<eventWithTime>,
    ) => listenerHandler | undefined;
    addCustomEvent<T>(tag: string, payload: T): void;
  };
  emit: (e: eventWithTime) => undefined;
  snapshots: eventWithTime[];
}

async function injectRecordScript(frame: puppeteer.Frame) {
  await frame.addScriptTag({
    path: path.resolve(__dirname, '../../dist/rrweb.js'),
  });
  await frame.evaluate(() => {
    ((window as unknown) as IWindow).snapshots = [];
    const { record } = ((window as unknown) as IWindow).rrweb;
    record({
      recordCrossOriginIframes: true,
      recordCanvas: true,
      emit(event) {
        ((window as unknown) as IWindow).snapshots.push(event);
        ((window as unknown) as IWindow).emit(event);
      },
    });
  });

  for (const child of frame.childFrames()) {
    await injectRecordScript(child);
  }
}

const setup = function (this: ISuite, content: string): ISuite {
  const ctx = {} as ISuite;

  beforeAll(async () => {
    ctx.browser = await launchPuppeteer();
    ctx.server = await startServer();
    ctx.serverURL = getServerURL(ctx.server);
    // ctx.serverB = await startServer();
    // ctx.serverBURL = getServerURL(ctx.serverB);

    const bundlePath = path.resolve(__dirname, '../../dist/rrweb.js');
    ctx.code = fs.readFileSync(bundlePath, 'utf8');
  });

  beforeEach(async () => {
    ctx.page = await ctx.browser.newPage();
    await ctx.page.goto('about:blank');
    await ctx.page.setContent(
      content.replace(/\{SERVER_URL\}/g, ctx.serverURL),
    );
    // await ctx.page.evaluate(ctx.code);
    ctx.events = [];
    await ctx.page.exposeFunction('emit', (e: eventWithTime) => {
      if (e.type === EventType.DomContentLoaded || e.type === EventType.Load) {
        return;
      }
      ctx.events.push(e);
    });

    ctx.page.on('console', (msg) => console.log('PAGE LOG:', msg.text()));
    await injectRecordScript(ctx.page.mainFrame());
  });

  afterEach(async () => {
    await ctx.page.close();
  });

  afterAll(async () => {
    await ctx.browser.close();
    ctx.server.close();
    // ctx.serverB.close();
  });

  return ctx;
};

describe('cross origin iframes', function (this: ISuite) {
  jest.setTimeout(100_000);

  describe('form.html', function (this: ISuite) {
    const ctx: ISuite = setup.call(
      this,
      `
        <!DOCTYPE html>
        <html>
          <body>
            <iframe src="{SERVER_URL}/html/form.html" style="width: 400px; height: 400px;"></iframe>
          </body>
        </html>
      `,
    );

    it("won't emit events if it isn't in the top level iframe", async () => {
      const el = (await ctx.page.$(
        'body > iframe',
      )) as puppeteer.ElementHandle<Element>;

      const frame = await el.contentFrame();
      const events = await frame?.evaluate(
        () => ((window as unknown) as IWindow).snapshots,
      );
      expect(events).toMatchObject([]);
    });

    it('will emit events if it is in the top level iframe', async () => {
      const events = await ctx.page.evaluate(
        () => ((window as unknown) as IWindow).snapshots,
      );
      expect(events.length).not.toBe(0);
    });

    it('should emit contents of iframe', async () => {
      const events = await ctx.page.evaluate(
        () => ((window as unknown) as IWindow).snapshots,
      );
      await waitForRAF(ctx.page);
      // two events from main frame, and two from iframe
      expect(events.length).toBe(4);
    });

    it('should emit full snapshot event from iframe as mutation event', async () => {
      const events = await ctx.page.evaluate(
        () => ((window as unknown) as IWindow).snapshots,
      );
      await waitForRAF(ctx.page);
      // two events from main frame, and two from iframe
      expect(events[events.length - 1]).toMatchObject({
        type: EventType.IncrementalSnapshot,
        data: {
          source: IncrementalSource.Mutation,
          adds: [
            {
              parentId: expect.any(Number),
              node: {
                id: expect.any(Number),
              },
            },
          ],
        },
      });
    });

    it('should use unique id for child of iframes', async () => {
      const events: eventWithTime[] = await ctx.page.evaluate(
        () => ((window as unknown) as IWindow).snapshots,
      );
      await waitForRAF(ctx.page);
      expect(
        (events[events.length - 1].data as mutationData).adds[0].node.id,
      ).not.toBe(1);
    });

    it('should replace the existing DOM nodes on iframe navigation with `isAttachIframe`', async () => {
      await ctx.page.evaluate((url) => {
        const iframe = document.querySelector('iframe') as HTMLIFrameElement;
        iframe.src = `${url}/html/form.html?2`;
      }, ctx.serverURL);
      await waitForRAF(ctx.page); // loads iframe

      await injectRecordScript(ctx.page.mainFrame().childFrames()[0]); // injects script into new iframe

      const events: eventWithTime[] = await ctx.page.evaluate(
        () => ((window as unknown) as IWindow).snapshots,
      );
      expect(
        (events[events.length - 1].data as mutationData).removes,
      ).toMatchObject([]);
      expect(
        (events[events.length - 1].data as mutationData).isAttachIframe,
      ).toBeTruthy();
    });

    it('should map input events correctly', async () => {
      const frame = ctx.page.mainFrame().childFrames()[0];
      await frame.type('input[type="text"]', 'test');
      await frame.click('input[type="radio"]');
      await frame.click('input[type="checkbox"]');
      await frame.type('input[type="password"]', 'password');
      await frame.type('textarea', 'textarea test');
      await frame.select('select', '1');

      const snapshots = (await ctx.page.evaluate(
        'window.snapshots',
      )) as eventWithTime[];
      assertSnapshot(snapshots);
    });

    it('should map scroll events correctly', async () => {
      // force scrollbars in iframe
      ctx.page.evaluate(() => {
        const iframe = document.querySelector('iframe') as HTMLIFrameElement;
        iframe.style.width = '50px';
        iframe.style.height = '50px';
      });

      await waitForRAF(ctx.page);
      const frame = ctx.page.mainFrame().childFrames()[0];

      // scroll a little
      frame.evaluate(() => {
        window.scrollTo(0, 10);
      });
      await waitForRAF(ctx.page);

      const snapshots = (await ctx.page.evaluate(
        'window.snapshots',
      )) as eventWithTime[];
      assertSnapshot(snapshots);
    });
  });

  describe('move-node.html', function (this: ISuite) {
    const ctx: ISuite = setup.call(
      this,
      `
        <!DOCTYPE html>
        <html>
          <body>
            <iframe src="{SERVER_URL}/html/move-node.html"></iframe>
          </body>
        </html>
      `,
    );

    it('should record DOM node movement', async () => {
      const frame = ctx.page.mainFrame().childFrames()[0];
      await frame.evaluate(() => {
        const div = document.createElement('div');
        const span = document.querySelector('span')!;
        document.body.appendChild(div);
        div.appendChild(span);
      });
      const snapshots = (await ctx.page.evaluate(
        'window.snapshots',
      )) as eventWithTime[];
      assertSnapshot(snapshots);
    });

    it('should record DOM node removal', async () => {
      const frame = ctx.page.mainFrame().childFrames()[0];
      await frame.evaluate(() => {
        const span = document.querySelector('span')!;
        span.remove();
      });
      const snapshots = (await ctx.page.evaluate(
        'window.snapshots',
      )) as eventWithTime[];
      assertSnapshot(snapshots);
    });

    it('should record DOM attribute changes', async () => {
      const frame = ctx.page.mainFrame().childFrames()[0];
      await frame.evaluate(() => {
        const span = document.querySelector('span')!;
        span.className = 'added-class-name';
      });
      const snapshots = (await ctx.page.evaluate(
        'window.snapshots',
      )) as eventWithTime[];
      assertSnapshot(snapshots);
    });

    it('should record DOM text changes', async () => {
      const frame = ctx.page.mainFrame().childFrames()[0];
      await frame.evaluate(() => {
        const b = document.querySelector('b')!;
        b.childNodes[0].textContent = 'replaced text';
      });
      const snapshots = (await ctx.page.evaluate(
        'window.snapshots',
      )) as eventWithTime[];
      assertSnapshot(snapshots);
    });

    it('should record canvas elements', async () => {
      const frame = ctx.page.mainFrame().childFrames()[0];
      await frame.evaluate(() => {
        var canvas = document.createElement('canvas');
        var gl = canvas.getContext('webgl')!;
        var program = gl.createProgram()!;
        gl.linkProgram(program);
        gl.clear(gl.COLOR_BUFFER_BIT);
        document.body.appendChild(canvas);
      });
      await waitForRAF(ctx.page);
      await waitForRAF(ctx.page);
      const snapshots = (await ctx.page.evaluate(
        'window.snapshots',
      )) as eventWithTime[];
      assertSnapshot(snapshots);
    });
  });
});

describe('same origin iframes', function (this: ISuite) {
  jest.setTimeout(100_000);

  const ctx: ISuite = setup.call(
    this,
    `
      <!DOCTYPE html>
      <html>
        <body>
          <iframe src="about:blank"></iframe>
        </body>
      </html>
    `,
  );

  it('should emit contents of iframe once', async () => {
    const events = await ctx.page.evaluate(
      () => ((window as unknown) as IWindow).snapshots,
    );
    await waitForRAF(ctx.page);
    // two events from main frame, and two from iframe
    expect(events.length).toBe(4);
  });
});
