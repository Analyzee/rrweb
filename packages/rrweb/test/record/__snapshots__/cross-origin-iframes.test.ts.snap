// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cross origin iframes should map input events correctly 1`] = `
"[
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 1920,
      \\"height\\": 1080
    }
  },
  {
    \\"type\\": 2,
    \\"data\\": {
      \\"node\\": {
        \\"type\\": 0,
        \\"childNodes\\": [
          {
            \\"type\\": 1,
            \\"name\\": \\"html\\",
            \\"publicId\\": \\"\\",
            \\"systemId\\": \\"\\",
            \\"id\\": 2
          },
          {
            \\"type\\": 2,
            \\"tagName\\": \\"html\\",
            \\"attributes\\": {},
            \\"childNodes\\": [
              {
                \\"type\\": 2,
                \\"tagName\\": \\"head\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"script\\",
                    \\"attributes\\": {
                      \\"type\\": \\"\\"
                    },
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"SCRIPT_PLACEHOLDER\\",
                        \\"id\\": 6
                      }
                    ],
                    \\"id\\": 5
                  }
                ],
                \\"id\\": 4
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"body\\",
                \\"attributes\\": {},
                \\"childNodes\\": [
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n          \\",
                    \\"id\\": 8
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"iframe\\",
                    \\"attributes\\": {
                      \\"rr_src\\": \\"http://localhost:3030/html/form.html\\"
                    },
                    \\"childNodes\\": [],
                    \\"id\\": 9
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n        \\\\n      \\\\n    \\",
                    \\"id\\": 10
                  }
                ],
                \\"id\\": 7
              }
            ],
            \\"id\\": 3
          }
        ],
        \\"id\\": 1
      },
      \\"initialOffset\\": {
        \\"left\\": 0,
        \\"top\\": 0
      }
    }
  },
  {
    \\"type\\": 4,
    \\"data\\": {
      \\"href\\": \\"about:blank\\",
      \\"width\\": 300,
      \\"height\\": 150
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 0,
      \\"adds\\": [
        {
          \\"parentId\\": 9,
          \\"nextId\\": null,
          \\"node\\": {
            \\"type\\": 0,
            \\"childNodes\\": [
              {
                \\"type\\": 1,
                \\"name\\": \\"html\\",
                \\"publicId\\": \\"\\",
                \\"systemId\\": \\"\\",
                \\"id\\": 12
              },
              {
                \\"type\\": 2,
                \\"tagName\\": \\"html\\",
                \\"attributes\\": {
                  \\"lang\\": \\"en\\"
                },
                \\"childNodes\\": [
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"head\\",
                    \\"attributes\\": {},
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"\\\\n    \\",
                        \\"id\\": 15
                      },
                      {
                        \\"type\\": 2,
                        \\"tagName\\": \\"meta\\",
                        \\"attributes\\": {
                          \\"charset\\": \\"UTF-8\\"
                        },
                        \\"childNodes\\": [],
                        \\"id\\": 16
                      },
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"\\\\n    \\",
                        \\"id\\": 17
                      },
                      {
                        \\"type\\": 2,
                        \\"tagName\\": \\"meta\\",
                        \\"attributes\\": {
                          \\"name\\": \\"viewport\\",
                          \\"content\\": \\"width=device-width, initial-scale=1.0\\"
                        },
                        \\"childNodes\\": [],
                        \\"id\\": 18
                      },
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"\\\\n    \\",
                        \\"id\\": 19
                      },
                      {
                        \\"type\\": 2,
                        \\"tagName\\": \\"meta\\",
                        \\"attributes\\": {
                          \\"http-equiv\\": \\"X-UA-Compatible\\",
                          \\"content\\": \\"ie=edge\\"
                        },
                        \\"childNodes\\": [],
                        \\"id\\": 20
                      },
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"\\\\n    \\",
                        \\"id\\": 21
                      },
                      {
                        \\"type\\": 2,
                        \\"tagName\\": \\"title\\",
                        \\"attributes\\": {},
                        \\"childNodes\\": [
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"form fields\\",
                            \\"id\\": 23
                          }
                        ],
                        \\"id\\": 22
                      },
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"\\\\n  \\",
                        \\"id\\": 24
                      },
                      {
                        \\"type\\": 2,
                        \\"tagName\\": \\"script\\",
                        \\"attributes\\": {
                          \\"type\\": \\"\\"
                        },
                        \\"childNodes\\": [
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"SCRIPT_PLACEHOLDER\\",
                            \\"id\\": 26
                          }
                        ],
                        \\"id\\": 25
                      }
                    ],
                    \\"id\\": 14
                  },
                  {
                    \\"type\\": 3,
                    \\"textContent\\": \\"\\\\n\\\\n  \\",
                    \\"id\\": 27
                  },
                  {
                    \\"type\\": 2,
                    \\"tagName\\": \\"body\\",
                    \\"attributes\\": {},
                    \\"childNodes\\": [
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"\\\\n    \\",
                        \\"id\\": 29
                      },
                      {
                        \\"type\\": 2,
                        \\"tagName\\": \\"form\\",
                        \\"attributes\\": {},
                        \\"childNodes\\": [
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n      \\",
                            \\"id\\": 31
                          },
                          {
                            \\"type\\": 2,
                            \\"tagName\\": \\"label\\",
                            \\"attributes\\": {
                              \\"for\\": \\"text\\"
                            },
                            \\"childNodes\\": [
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n        \\",
                                \\"id\\": 33
                              },
                              {
                                \\"type\\": 2,
                                \\"tagName\\": \\"input\\",
                                \\"attributes\\": {
                                  \\"type\\": \\"text\\"
                                },
                                \\"childNodes\\": [],
                                \\"id\\": 34
                              },
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n      \\",
                                \\"id\\": 35
                              }
                            ],
                            \\"id\\": 32
                          },
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n      \\",
                            \\"id\\": 36
                          },
                          {
                            \\"type\\": 2,
                            \\"tagName\\": \\"label\\",
                            \\"attributes\\": {},
                            \\"childNodes\\": [
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n        \\",
                                \\"id\\": 38
                              },
                              {
                                \\"type\\": 2,
                                \\"tagName\\": \\"input\\",
                                \\"attributes\\": {
                                  \\"type\\": \\"radio\\",
                                  \\"name\\": \\"toggle\\",
                                  \\"value\\": \\"on\\"
                                },
                                \\"childNodes\\": [],
                                \\"id\\": 39
                              },
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n      \\",
                                \\"id\\": 40
                              }
                            ],
                            \\"id\\": 37
                          },
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n      \\",
                            \\"id\\": 41
                          },
                          {
                            \\"type\\": 2,
                            \\"tagName\\": \\"label\\",
                            \\"attributes\\": {},
                            \\"childNodes\\": [
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n        \\",
                                \\"id\\": 43
                              },
                              {
                                \\"type\\": 2,
                                \\"tagName\\": \\"input\\",
                                \\"attributes\\": {
                                  \\"type\\": \\"radio\\",
                                  \\"name\\": \\"toggle\\",
                                  \\"value\\": \\"off\\",
                                  \\"checked\\": true
                                },
                                \\"childNodes\\": [],
                                \\"id\\": 44
                              },
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n      \\",
                                \\"id\\": 45
                              }
                            ],
                            \\"id\\": 42
                          },
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n      \\",
                            \\"id\\": 46
                          },
                          {
                            \\"type\\": 2,
                            \\"tagName\\": \\"label\\",
                            \\"attributes\\": {
                              \\"for\\": \\"checkbox\\"
                            },
                            \\"childNodes\\": [
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n        \\",
                                \\"id\\": 48
                              },
                              {
                                \\"type\\": 2,
                                \\"tagName\\": \\"input\\",
                                \\"attributes\\": {
                                  \\"type\\": \\"checkbox\\"
                                },
                                \\"childNodes\\": [],
                                \\"id\\": 49
                              },
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n      \\",
                                \\"id\\": 50
                              }
                            ],
                            \\"id\\": 47
                          },
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n      \\",
                            \\"id\\": 51
                          },
                          {
                            \\"type\\": 2,
                            \\"tagName\\": \\"label\\",
                            \\"attributes\\": {
                              \\"for\\": \\"textarea\\"
                            },
                            \\"childNodes\\": [
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n        \\",
                                \\"id\\": 53
                              },
                              {
                                \\"type\\": 2,
                                \\"tagName\\": \\"textarea\\",
                                \\"attributes\\": {
                                  \\"name\\": \\"\\",
                                  \\"id\\": \\"\\",
                                  \\"cols\\": \\"30\\",
                                  \\"rows\\": \\"10\\"
                                },
                                \\"childNodes\\": [],
                                \\"id\\": 54
                              },
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n      \\",
                                \\"id\\": 55
                              }
                            ],
                            \\"id\\": 52
                          },
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n      \\",
                            \\"id\\": 56
                          },
                          {
                            \\"type\\": 2,
                            \\"tagName\\": \\"label\\",
                            \\"attributes\\": {
                              \\"for\\": \\"select\\"
                            },
                            \\"childNodes\\": [
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n        \\",
                                \\"id\\": 58
                              },
                              {
                                \\"type\\": 2,
                                \\"tagName\\": \\"select\\",
                                \\"attributes\\": {
                                  \\"name\\": \\"\\",
                                  \\"id\\": \\"\\",
                                  \\"value\\": \\"1\\"
                                },
                                \\"childNodes\\": [
                                  {
                                    \\"type\\": 3,
                                    \\"textContent\\": \\"\\\\n          \\",
                                    \\"id\\": 60
                                  },
                                  {
                                    \\"type\\": 2,
                                    \\"tagName\\": \\"option\\",
                                    \\"attributes\\": {
                                      \\"value\\": \\"1\\",
                                      \\"selected\\": true
                                    },
                                    \\"childNodes\\": [
                                      {
                                        \\"type\\": 3,
                                        \\"textContent\\": \\"1\\",
                                        \\"id\\": 62
                                      }
                                    ],
                                    \\"id\\": 61
                                  },
                                  {
                                    \\"type\\": 3,
                                    \\"textContent\\": \\"\\\\n          \\",
                                    \\"id\\": 63
                                  },
                                  {
                                    \\"type\\": 2,
                                    \\"tagName\\": \\"option\\",
                                    \\"attributes\\": {
                                      \\"value\\": \\"2\\"
                                    },
                                    \\"childNodes\\": [
                                      {
                                        \\"type\\": 3,
                                        \\"textContent\\": \\"2\\",
                                        \\"id\\": 65
                                      }
                                    ],
                                    \\"id\\": 64
                                  },
                                  {
                                    \\"type\\": 3,
                                    \\"textContent\\": \\"\\\\n        \\",
                                    \\"id\\": 66
                                  }
                                ],
                                \\"id\\": 59
                              },
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n      \\",
                                \\"id\\": 67
                              }
                            ],
                            \\"id\\": 57
                          },
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n      \\",
                            \\"id\\": 68
                          },
                          {
                            \\"type\\": 2,
                            \\"tagName\\": \\"label\\",
                            \\"attributes\\": {
                              \\"for\\": \\"password\\"
                            },
                            \\"childNodes\\": [
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n        \\",
                                \\"id\\": 70
                              },
                              {
                                \\"type\\": 2,
                                \\"tagName\\": \\"input\\",
                                \\"attributes\\": {
                                  \\"type\\": \\"password\\"
                                },
                                \\"childNodes\\": [],
                                \\"id\\": 71
                              },
                              {
                                \\"type\\": 3,
                                \\"textContent\\": \\"\\\\n      \\",
                                \\"id\\": 72
                              }
                            ],
                            \\"id\\": 69
                          },
                          {
                            \\"type\\": 3,
                            \\"textContent\\": \\"\\\\n    \\",
                            \\"id\\": 73
                          }
                        ],
                        \\"id\\": 30
                      },
                      {
                        \\"type\\": 3,
                        \\"textContent\\": \\"\\\\n  \\\\n\\\\n\\",
                        \\"id\\": 74
                      }
                    ],
                    \\"id\\": 28
                  }
                ],
                \\"id\\": 13
              }
            ],
            \\"id\\": 11
          }
        }
      ],
      \\"removes\\": [],
      \\"texts\\": [],
      \\"attributes\\": [],
      \\"isAttachIframe\\": true
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 5,
      \\"id\\": 24
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"t\\",
      \\"isChecked\\": false,
      \\"id\\": 34
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"te\\",
      \\"isChecked\\": false,
      \\"id\\": 34
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"tes\\",
      \\"isChecked\\": false,
      \\"id\\": 34
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"test\\",
      \\"isChecked\\": false,
      \\"id\\": 34
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 1,
      \\"id\\": 29
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 6,
      \\"id\\": 24
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 5,
      \\"id\\": 29
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 0,
      \\"id\\": 29
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 2,
      \\"id\\": 29
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"on\\",
      \\"isChecked\\": true,
      \\"id\\": 39
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"off\\",
      \\"isChecked\\": false,
      \\"id\\": 44
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 1,
      \\"id\\": 39
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 6,
      \\"id\\": 29
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 5,
      \\"id\\": 39
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 0,
      \\"id\\": 39
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 2,
      \\"id\\": 39
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"on\\",
      \\"isChecked\\": true,
      \\"id\\": 49
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 6,
      \\"id\\": 39
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 5,
      \\"id\\": 61
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 3,
      \\"id\\": 1,
      \\"x\\": 0,
      \\"y\\": 69
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"*\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"**\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"***\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"****\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"*****\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"******\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"*******\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"********\\",
      \\"isChecked\\": false,
      \\"id\\": 71
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 6,
      \\"id\\": 61
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 2,
      \\"type\\": 5,
      \\"id\\": 44
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"t\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"te\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"tex\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"text\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"texta\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textar\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textare\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textarea\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textarea \\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textarea t\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textarea te\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textarea tes\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"textarea test\\",
      \\"isChecked\\": false,
      \\"id\\": 54
    }
  },
  {
    \\"type\\": 3,
    \\"data\\": {
      \\"source\\": 5,
      \\"text\\": \\"1\\",
      \\"isChecked\\": false,
      \\"id\\": 59
    }
  }
]"
`;
